pysph/sph/surface_tension.py

```diff
@@ -13,11 +13,240 @@
+        d_au[d_idx] += (DWIJ[0]*f00 + DWIJ[1]*f01 + DWIJ[2]*f02)/d_m[d_idx]
+        d_av[d_idx] += (DWIJ[0]*f10 + DWIJ[1]*f11 + DWIJ[2]*f12)/d_m[d_idx]
+        d_aw[d_idx] += (DWIJ[0]*f20 + DWIJ[1]*f21 + DWIJ[2]*f22)/d_m[d_idx]
+
+
+class ConstructStressMatrix(Equation):
+
+    def __init__(self, dest, sources, sigma, d=2):
+        self.sigma = sigma
+        self.d = d
+        super(ConstructStressMatrix, self).__init__(dest, sources)
+
+    def initialize(self, d_pi00, d_pi01, d_pi02, d_pi10, d_pi11, d_pi12,
+                   d_pi20, d_pi21, d_pi22, d_cx, d_cy, d_cz, d_idx, d_N):
+        mod_gradc2 = d_cx[d_idx]*d_cx[d_idx] + d_cy[d_idx]*d_cy[d_idx] + \
+            d_cz[d_idx]*d_cz[d_idx]
+        mod_gradc = sqrt(mod_gradc2)
+        d_N[d_idx] = 0.0
+        if mod_gradc > 1e-14:
+            factor = self.sigma/mod_gradc
+            d_pi00[d_idx] = (-d_cx[d_idx]*d_cx[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_pi01[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi02[d_idx] = -factor*d_cx[d_idx]*d_cz[d_idx]
+            d_pi10[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi11[d_idx] = (-d_cy[d_idx]*d_cy[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_pi12[d_idx] = -factor*d_cy[d_idx]*d_cz[d_idx]
+            d_pi20[d_idx] = -factor*d_cx[d_idx]*d_cz[d_idx]
+            d_pi21[d_idx] = -factor*d_cy[d_idx]*d_cz[d_idx]
+            d_pi22[d_idx] = (-d_cz[d_idx]*d_cz[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_N[d_idx] = 1.0
+        else:
+            d_pi00[d_idx] = 0.0
+            d_pi01[d_idx] = 0.0
+            d_pi02[d_idx] = 0.0
+            d_pi10[d_idx] = 0.0
+            d_pi11[d_idx] = 0.0
+            d_pi12[d_idx] = 0.0
+            d_pi20[d_idx] = 0.0
+            d_pi21[d_idx] = 0.0
+            d_pi22[d_idx] = 0.0
+
+
+class ColorGradientAdami(Equation):
+
+    def initialize(self, d_idx, d_cx, d_cy, d_cz):
+        d_cx[d_idx] = 0.0
+        d_cy[d_idx] = 0.0
+        d_cz[d_idx] = 0.0
+
+    def loop(self, d_idx, d_cx, d_cy, d_cz, d_V, s_V, d_color, s_color, DWIJ,
+             s_idx):
+        c_i = d_color[d_idx]/(d_V[d_idx]*d_V[d_idx])
+        c_j = s_color[s_idx]/(s_V[s_idx]*s_V[s_idx])
+        factor = d_V[d_idx]*(c_i + c_j)
+        d_cx[d_idx] += factor*DWIJ[0]
+        d_cy[d_idx] += factor*DWIJ[1]
+        d_cz[d_idx] += factor*DWIJ[2]
+
+
+class MomentumEquationViscosityAdami(Equation):
+
+    def initialize(self, d_au, d_av, d_aw, d_idx):
+        d_au[d_idx] = 0.0
+        d_av[d_idx] = 0.0
+        d_aw[d_idx] = 0.0
+
+    def loop(self, d_idx, d_V, d_au, d_av, d_aw, s_V, d_p, s_p, DWIJ, s_idx,
+             d_m, R2IJ, XIJ, EPS, VIJ, d_nu, s_nu):
+        factor = 2.0*d_nu[d_idx]*s_nu[s_idx]/(d_nu[d_idx] + s_nu[s_idx])
+        V_i = 1/(d_V[d_idx]*d_V[d_idx])
+        V_j = 1/(s_V[s_idx]*s_V[s_idx])
+        dwijdotrij = (DWIJ[0]*XIJ[0]+DWIJ[1]*XIJ[1]+DWIJ[2]*XIJ[2])
```

Comment:
Please leave space between `+` and `-` operators for readability.

Reply:
Done. Please check now.